equinox/nn/_embedding.py

```diff
@@ -89,3 +96,313 @@ def __call__(
                 "`eqx.nn.Embedding()(x)` should be called with a scalar index `x`. "
                 "Use `jax.vmap` if you would like to index with multiple values."
             )
+
+
+class RotaryPositionalEmbedding(Module, strict=True):
+    """
+    A rotary positional encoding module, as described in the paper
+    "RoFormer: Enhanced Transformer with Rotary Position Embedding". While
+    this module can be used in any context, it is particularly useful for
+    providing positional information to transformer models.
+
+    !!! example
+        The following example demonstrates how to use `RotaryPositionalEmbedding` in
+        a simple transformer model.
+        ```python
+
+        class TransformerBlock(eqx.nn.StatefulLayer):
+            ...
+            key_rope_embeddings: RotaryPositionalEmbedding
+            query_rope_embeddings: RotaryPositionalEmbedding
+
+            def __init__(...):
+                ...
+                self.query_rope_embeddings = RotaryPositionalEmbedding(
+                    embedding_size=n_embd
+                )
+                self.key_rope_embeddings = RotaryPositionalEmbedding(
+                    embedding_size=n_embd
+                )
+                ...
+
+            def __call__(...):
+                def process_heads(query_heads, key_heads, value_heads):
+                    query_heads = jax.vmap(self.query_rope_embeddings,
+                                           in_axes=1,
+                                           out_axes=1)(query_heads)
+                    key_heads = jax.vmap(self.key_rope_embeddings,
+                                         in_axes=1,
+                                         out_axes=1)(key_heads)
+
+                    return query_heads, key_heads, value_heads
+
+                mha_output = self.mha_attention(
+                    process_heads=process_heads,
+                    query=jax.vmap(self.rms_norm)(x),
```

Comment:
I'd probably remove the explicit `query/key/value/mask=...` inputs here. They don't directly affect the purpose of this example so they're just visual noise.
(They're also suggestive of doing a few particular things, like self-attention and using RMSNorm, which may not be needed in general.)